Added torch.linalg.matrix_power

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53538 Deprecate torch.matrix_power in favor of torch.linalg.matrix_power
* **#52608 Added torch.linalg.matrix_power**

**TODO**

- [x] Add OpInfo
- [x] Update documentation
- [x] Add more tests and compare against NumPy

Differential Revision: [D27261532](https://our.internmc.facebook.com/intern/diff/D27261532)

heitorschueroff committed:
Deprecated torch.matrix_power in favor of torch.linalg.matrix_power

[ghstack-poisoned]

heitorschueroff committed:
Update on "Deprecated torch.matrix_power in favor of torch.linalg.matrix_power"


**TODO**

- [x] Add OpInfo
- [x] Update documentation
- [ ] Add more tests and compare against NumPy
- [ ] Benchmark against NumPy

[ghstack-poisoned]

heitorschueroff committed:
Update on "Deprecated torch.matrix_power in favor of torch.linalg.matrix_power"


**TODO**

- [x] Add OpInfo
- [x] Update documentation
- [x] Add more tests and compare against NumPy
- [ ] Benchmark against NumPy

[ghstack-poisoned]

heitorschueroff committed:
Update on "Added torch.linalg.matrix_power"


**TODO**

- [x] Add OpInfo
- [x] Update documentation
- [x] Add more tests and compare against NumPy
- [ ] Benchmark against NumPy

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -195,6 +195,59 @@ Tensor pinverse(const Tensor& self, double rcond) {
+  checkSameDevice("matrix_power", result, self);
+  checkLinalgCompatibleDtype("matrix_power", result, self);
+  at::native::resize_output(result, self.sizes());
+  result.copy_(at::native::linalg_matrix_power(self, n));
```
If we have to keep this pattern let's use this opportunity to implement a safe copy function. Can we not fuse this copy, however?

heitorschueroff replied:
Implemented out version.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -195,6 +195,59 @@ Tensor pinverse(const Tensor& self, double rcond) {
+
+  auto a = n < 0 ? self.inverse() : self.clone(at::MemoryFormat::Contiguous);
+  n = std::abs(n);
+
```
Comment for this block (fast paths for common special cases)

mruberry commented on test/test_linalg.py:
```diff
@@ -5346,48 +5346,39 @@ def run_test(M):
+    def test_matrix_power_non_negative(self, device, dtype):
+        def check(*size, discontiguous=False):
+            t = make_tensor(size, device, dtype, discontiguous=discontiguous)
+            for n in range(5):
```
How about range(8) so some odd general cases are tested?

mruberry commented on test/test_linalg.py:
```diff
@@ -5346,48 +5346,39 @@ def run_test(M):
-        run_test(M, sign=-1)
+        def check(*size):
+            t = random_fullrank_matrix_distinct_singular_value(*size, dtype=dtype, device=device)
+            for n in range(-4, 0):
```
What about expanding the range of the previous test to be -7 to 7?

heitorschueroff replied:
Sure, I can combine both tests into one and expand the range.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -367,6 +367,49 @@
+Raises the square matrix :attr:`input`, or each square matrix in a batched
+:attr:`input`, to the integer power :attr:`n`.
+
+If :attr:`n` is 0, the identity matrix of the same shape as :attr:`input` is returned.
```
identity matrix (or batch of identity matrices)

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -367,6 +367,49 @@
+:attr:`input`, to the integer power :attr:`n`.
+
+If :attr:`n` is 0, the identity matrix of the same shape as :attr:`input` is returned.
+If :attr:`n` is negative, the inverse of :attr:`input` (if invertible) is computed and
```
"the inverse of input" -> "the inverse of each matrix"

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1811,6 +1833,19 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
+           sample_inputs_func=sample_inputs_linalg_matrix_power,
+           aten_name='linalg_matrix_power',
+           skips=(
+               # RuntimeError: input->type()->kind() == TypeKind::OptionalTypeINTERNAL ASSERT FAILED
```
Would you elaborate on this? If this issue isn't addressed in this PR let's file an issue and post its url in this comment.

heitorschueroff replied:
Removed the skip, tests seems to pass now.

mruberry reviewed (commented):
This looks good to me except for the out= behavior, which would be nice to address (and maybe providing the helper function for safe copying we've discussed before).

The ideep/tensorpipe changes should be reverted and the PR rebased past the mypy failures in its base.

The plan is for a follow-up PR to alias torch.matrix_power to torch.linalg.matrix_power, right?

@IvanYashchuk would you like to take a look?

IvanYashchuk commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -195,6 +195,59 @@ Tensor pinverse(const Tensor& self, double rcond) {
+      "matrix_power: expected input tensor of floating or complex type but got ",
+      self.scalar_type());
+
+  auto a = n < 0 ? self.inverse() : self.clone(at::MemoryFormat::Contiguous);
```
Let's start using new `at::linalg_inv` instead of `at::inverse/.inverse` for the internal code as well.

IvanYashchuk commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1811,6 +1833,19 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
+           dtypes=floating_and_complex_types(),
+           test_inplace_grad=False,
+           supports_tensor_out=True,
+           decorators=[skipCUDAIfNoMagma, skipCPUIfNoLapack],
```
Is matrix inverse working for ROCm? If not it might also need `skipCUDAIfROCm`.
`skipCUDAIfNoMagma` should be `skipCUDAIfNoMagmaAndNoCusolver` because inverse can use either of those.

heitorschueroff replied:
Here are the test durations

```
11.87s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_linalg_matrix_power_cuda_complex128
5.83s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_linalg_matrix_power_cpu_complex128
3.13s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_linalg_matrix_power_cuda_float64
2.73s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_linalg_matrix_power_cuda_complex128
1.79s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_matrix_power_cuda_complex128
1.71s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_linalg_matrix_power_cpu_float64
1.36s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_matrix_power_cuda_complex128
1.34s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_matrix_power_cuda_complex64
1.24s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_matrix_power_cuda_float32
1.23s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_matrix_power_cuda_float64
1.05s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_linalg_matrix_power_cpu_complex128
1.01s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_linalg_matrix_power_cpu_complex64
0.96s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_linalg_matrix_power_cpu_float32
0.95s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_linalg_matrix_power_cpu_float64
0.68s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_linalg_matrix_power_cpu_complex128
0.60s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_matrix_power_cuda_float64
0.24s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_linalg_matrix_power_cpu_float64
0.13s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_eager_linalg_matrix_power_cuda_float64
0.13s call     test/test_ops.py::TestOpInfoCPU::test_supported_dtypes_linalg_matrix_power_cpu_complex128
0.13s call     test/test_ops.py::TestCommonCUDA::test_out_linalg_matrix_power_cuda_complex128
0.12s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_eager_linalg_matrix_power_cuda_complex64
0.12s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_eager_linalg_matrix_power_cuda_complex128
0.11s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_linalg_matrix_power_cuda_float16
0.11s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_eager_linalg_matrix_power_cuda_float32
0.11s call     test/test_ops.py::TestCommonCUDA::test_out_linalg_matrix_power_cuda_complex64
0.11s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_linalg_matrix_power_cuda_bool
0.10s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_linalg_matrix_power_cuda_int64
0.10s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_linalg_matrix_power_cuda_complex64
0.10s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_linalg_matrix_power_cuda_float64
0.10s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_linalg_matrix_power_cuda_float32
0.10s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_linalg_matrix_power_cuda_bfloat16
0.10s call     test/test_ops.py::TestCommonCUDA::test_out_linalg_matrix_power_cuda_float64
0.10s call     test/test_ops.py::TestCommonCUDA::test_out_linalg_matrix_power_cuda_float32
0.10s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_linalg_matrix_power_cuda_int8
0.10s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_linalg_matrix_power_cuda_int32
0.10s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_linalg_matrix_power_cuda_int16
0.10s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_linalg_matrix_power_cuda_uint8
0.05s call     test/test_ops.py::TestGradientsCUDA::test_inplace_gradgrad_linalg_matrix_power_cuda_float64
0.05s call     test/test_ops.py::TestGradientsCUDA::test_inplace_grad_linalg_matrix_power_cuda_complex128
0.05s call     test/test_ops.py::TestGradientsCUDA::test_inplace_grad_linalg_matrix_power_cuda_float64
0.05s call     test/test_ops.py::TestGradientsCUDA::test_inplace_gradgrad_linalg_matrix_power_cuda_complex128
0.01s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_eager_linalg_matrix_power_cpu_complex64
0.01s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_eager_linalg_matrix_power_cpu_complex128
0.01s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_eager_linalg_matrix_power_cpu_float64
0.01s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_eager_linalg_matrix_power_cpu_float32
0.01s call     test/test_ops.py::TestOpInfoCPU::test_unsupported_dtypes_linalg_matrix_power_cpu_int32
0.01s call     test/test_ops.py::TestOpInfoCPU::test_unsupported_dtypes_linalg_matrix_power_cpu_int64
0.01s call     test/test_ops.py::TestOpInfoCPU::test_unsupported_dtypes_linalg_matrix_power_cpu_bool
0.01s call     test/test_ops.py::TestOpInfoCPU::test_unsupported_dtypes_linalg_matrix_power_cpu_uint8
0.01s call     test/test_ops.py::TestOpInfoCPU::test_unsupported_dtypes_linalg_matrix_power_cpu_int16
0.01s call     test/test_ops.py::TestOpInfoCPU::test_unsupported_dtypes_linalg_matrix_power_cpu_int8

(141 durations < 0.005s hidden.  Use -vv to show these durations.)
```

IvanYashchuk commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -358,6 +358,28 @@ def sample_inputs_tensor_split(op_info, device, dtype, requires_grad):
+        (1, 1),
+        (S, S),
+        (0, S, S),
+        (S, S, S),
```
`S` as a batch dimension slows down the gradgradchecks. Using 2 is better.

IvanYashchuk commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -358,6 +358,28 @@ def sample_inputs_tensor_split(op_info, device, dtype, requires_grad):
+            inputs.append(SampleInput(t, args=(n,)))
+
+    for n in torch.arange(-4, 0):
+        t = random_fullrank_matrix_distinct_singular_value(S, S, device=device, dtype=dtype)
```
I suggest to use 2 instead of S for batches.
```suggestion
        t = random_fullrank_matrix_distinct_singular_value(S, 2, device=device, dtype=dtype)
```

IvanYashchuk reviewed (commented):
Looks good to me!

@heitorschueroff could you also post the runtime for the OpInfo checks? (`--durations=0` option for pytest)

heitorschueroff committed:
Update on "Added torch.linalg.matrix_power"


**TODO**

- [x] Add OpInfo
- [x] Update documentation
- [x] Add more tests and compare against NumPy
- [ ] Benchmark against NumPy

[ghstack-poisoned]

heitorschueroff committed:
Update on "Added torch.linalg.matrix_power"


**TODO**

- [x] Add OpInfo
- [x] Update documentation
- [x] Add more tests and compare against NumPy
- [ ] Benchmark against NumPy

[ghstack-poisoned]

heitorschueroff committed:
Update on "Added torch.linalg.matrix_power"


**TODO**

- [x] Add OpInfo
- [x] Update documentation
- [x] Add more tests and compare against NumPy
- [ ] Benchmark against NumPy

[ghstack-poisoned]

heitorschueroff committed:
Update on "Added torch.linalg.matrix_power"


**TODO**

- [x] Add OpInfo
- [x] Update documentation
- [x] Add more tests and compare against NumPy
- [ ] Benchmark against NumPy

[ghstack-poisoned]

heitorschueroff commented:
> Looks good to me!
> 
> @heitorschueroff could you also post the runtime for the OpInfo checks? (`--durations=0` option for pytest)

```
======================================================================================================= slowest durations =======================================================================================================
11.75s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_linalg_matrix_power_cuda_complex128
5.45s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_linalg_matrix_power_cpu_complex128
3.07s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_linalg_matrix_power_cuda_float64
2.65s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_linalg_matrix_power_cuda_complex128
1.73s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_matrix_power_cuda_complex128
1.65s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_linalg_matrix_power_cpu_float64
1.32s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_matrix_power_cuda_complex128
1.32s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_matrix_power_cuda_complex64
1.20s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_matrix_power_cuda_float32
1.19s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_matrix_power_cuda_float64
1.04s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_linalg_matrix_power_cpu_complex128
1.00s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_linalg_matrix_power_cpu_complex64
0.95s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_linalg_matrix_power_cpu_float32
0.93s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_linalg_matrix_power_cpu_float64
0.65s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_linalg_matrix_power_cpu_complex128
0.56s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_matrix_power_cuda_float64
0.22s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_linalg_matrix_power_cpu_float64
0.13s call     test/test_ops.py::TestOpInfoCPU::test_supported_dtypes_linalg_matrix_power_cpu_complex128
0.11s call     test/test_ops.py::TestCommonCUDA::test_out_linalg_matrix_power_cuda_float32
0.11s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_linalg_matrix_power_cuda_complex64
0.10s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_eager_linalg_matrix_power_cuda_complex64
0.10s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_eager_linalg_matrix_power_cuda_complex128
0.10s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_eager_linalg_matrix_power_cuda_float32
0.09s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_eager_linalg_matrix_power_cuda_float64
0.09s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_linalg_matrix_power_cuda_bool
0.09s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_linalg_matrix_power_cuda_float64
0.08s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_linalg_matrix_power_cuda_float32
0.08s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_linalg_matrix_power_cuda_bfloat16
0.08s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_linalg_matrix_power_cuda_int16
0.08s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_linalg_matrix_power_cuda_uint8
0.08s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_linalg_matrix_power_cuda_int8
0.08s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_linalg_matrix_power_cuda_float16
0.08s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_linalg_matrix_power_cuda_int64
0.08s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_linalg_matrix_power_cuda_int32
0.05s call     test/test_ops.py::TestGradientsCUDA::test_inplace_grad_linalg_matrix_power_cuda_complex128
0.04s call     test/test_ops.py::TestGradientsCUDA::test_inplace_gradgrad_linalg_matrix_power_cuda_complex128
0.04s call     test/test_ops.py::TestGradientsCUDA::test_inplace_gradgrad_linalg_matrix_power_cuda_float64
0.04s call     test/test_ops.py::TestGradientsCUDA::test_inplace_grad_linalg_matrix_power_cuda_float64
0.01s call     test/test_ops.py::TestCommonCPU::test_out_linalg_matrix_power_cpu_float32
0.01s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_eager_linalg_matrix_power_cpu_complex64
0.01s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_eager_linalg_matrix_power_cpu_complex128
0.01s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_eager_linalg_matrix_power_cpu_float64
0.01s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_eager_linalg_matrix_power_cpu_float32
0.01s call     test/test_ops.py::TestOpInfoCPU::test_unsupported_dtypes_linalg_matrix_power_cpu_bool
0.01s call     test/test_ops.py::TestOpInfoCPU::test_unsupported_dtypes_linalg_matrix_power_cpu_int16
0.01s call     test/test_ops.py::TestOpInfoCPU::test_unsupported_dtypes_linalg_matrix_power_cpu_int64
0.01s call     test/test_ops.py::TestOpInfoCPU::test_unsupported_dtypes_linalg_matrix_power_cpu_int32
0.01s call     test/test_ops.py::TestOpInfoCPU::test_unsupported_dtypes_linalg_matrix_power_cpu_int8
0.01s call     test/test_ops.py::TestOpInfoCPU::test_unsupported_dtypes_linalg_matrix_power_cpu_uint8
0.01s call     test/test_ops.py::TestOpInfoCPU::test_unsupported_dtypes_linalg_matrix_power_cpu_bfloat16

(124 durations < 0.005s hidden.  Use -vv to show these durations.)
========================================================================================
```

heitorschueroff committed:
Update on "Added torch.linalg.matrix_power"


**TODO**

- [x] Add OpInfo
- [x] Update documentation
- [x] Add more tests and compare against NumPy
- [ ] Benchmark against NumPy

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -358,6 +358,28 @@ def sample_inputs_tensor_split(op_info, device, dtype, requires_grad):
+    # (<matrix_size>, (<batch_sizes, ...>))
+    test_sizes = [
+        (1, ()),
+        (S, ()),
```
Let's use 2 instead of S

heitorschueroff replied:
Ok, test duration came down to 22 seconds with this change

mruberry replied:
That should be (barely) OK for now.

mruberry reviewed (approved):
Nice work, @heitorschueroff!

I think the test failures are erroneous. You may want to rebase to verify the docs build failure isn't related to this PR.

heitorschueroff committed:
Update on "Added torch.linalg.matrix_power"


**TODO**

- [x] Add OpInfo
- [x] Update documentation
- [x] Add more tests and compare against NumPy
- [ ] Benchmark against NumPy

[ghstack-poisoned]

mruberry reviewed (commented):
This accidentally picked up some fbgemm and tensorpipe changes that should be reverted.

heitorschueroff committed:
Update on "Added torch.linalg.matrix_power"


**TODO**

- [x] Add OpInfo
- [x] Update documentation
- [x] Add more tests and compare against NumPy

[ghstack-poisoned]

heitorschueroff committed:
Update on "Added torch.linalg.matrix_power"


**TODO**

- [x] Add OpInfo
- [x] Update documentation
- [x] Add more tests and compare against NumPy

[ghstack-poisoned]

heitorschueroff committed:
Update on "Added torch.linalg.matrix_power"


**TODO**

- [x] Add OpInfo
- [x] Update documentation
- [x] Add more tests and compare against NumPy

[ghstack-poisoned]

heitorschueroff committed:
Update on "Added torch.linalg.matrix_power"


**TODO**

- [x] Add OpInfo
- [x] Update documentation
- [x] Add more tests and compare against NumPy

Differential Revision: [D27260630](https://our.internmc.facebook.com/intern/diff/D27260630)

[ghstack-poisoned]
